Initial, partial support for `externref`

fitzgen opened the pull request:
This is enough to get an `externref -> externref` identity function passing.

However, `externref`s that are dropped by compiled Wasm code are (safely) leaked. Follow up work will leverage cranelift's stack maps to resolve this issue.

fitzgen commented:
Alright, I was missing some C API changes, but I've added them, and CI is green now.

alexcrichton commented on crates/runtime/src/externref.rs:
```diff
@@ -0,0 +1,445 @@
+        data.decrement_ref_count();
+        if data.get_ref_count() == 0 {
+            // Drop our live reference to `data` before we drop it itself.
+            drop(data);
```
Since `data` isn't an owned value this doesn't actually do anything, right?

fitzgen replied:
Yeah, this is more of a doing-everything-by-the-book and crossing-out-ts-and-dotting-our-is sort of thing for avoiding UB, but it is maybe overly conservative.

alexcrichton commented on crates/runtime/src/externref.rs:
```diff
@@ -0,0 +1,445 @@
+                let alloc_size = value_and_padding_size + extern_data_size;
+
+                debug_assert!(Layout::from_size_align(alloc_size, alloc_align).is_ok());
+                Layout::from_size_align_unchecked(alloc_size, alloc_align)
```
This looks like it's sharing a lot of code with allocation below, could that be refactored out to a shared function?

alexcrichton commented on crates/runtime/src/externref.rs:
```diff
@@ -0,0 +1,445 @@
+                let extern_data_align = mem::align_of::<VMExternData>();
+
+                let value_and_padding_size = round_up_to_align(value_size, extern_data_align)
+                    .unwrap_or_else(|| unreachable!());
```
`.unwrap()`?

alexcrichton commented on crates/runtime/src/externref.rs:
```diff
@@ -0,0 +1,445 @@
+    #[inline(never)]
+    #[cold]
+    fn alloc_failure() -> ! {
+        panic!("VMExternRef allocation failure")
```
Perhaps defer to https://doc.rust-lang.org/std/alloc/fn.handle_alloc_error.html?

alexcrichton commented on crates/runtime/src/externref.rs:
```diff
@@ -0,0 +1,445 @@
+    }
+}
+
+impl PartialEq for VMExternRef {
```
This and traits below seem dangerous since I'd naively expect it to defer to the underlying type (which can't be done here b/c it's not generic). Could this be exposed as a free-function?

alexcrichton commented on crates/wasmtime/src/func.rs:
```diff
@@ -722,7 +727,10 @@ impl Func {
+    // This should be `pub(crate)` but the C API happens to need it in one
+    // place.
+    #[doc(hidden)]
+    pub fn store(&self) -> &Store {
```
Is this a place in the C API where the store should be passed in?

alexcrichton replied:
This seems like a fine API to expose though, especially because `Caller` below exposes it too

fitzgen replied:
This is needed for the `wasi_instance_bind_import` WASI C API function -- do you happen to know if this is part of the C API or part of wasi?

alexcrichton replied:
Ah I think that's just part of WASI, but this seems fine to add anyway

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -2,224 +2,179 @@
 
```
Looks like this file became executable?

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -2,224 +2,179 @@
     /// Creates a `Null` reference.
-    pub fn null() -> Self {
-        ExternRef::Null
+    pub fn null() -> Option<Self> {
```
This can probably just be deleted now, it doesn't seem to serve too much purpose any more

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -2,224 +2,179 @@
+        &*self.inner
+    }
+
+    /// just point to two different values that compare as equal).
```
out of date comment?

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -2,224 +2,179 @@
-        }
+    /// Returns the host information for this `externref`, if previously created
+    /// with `set_host_info`.
+    pub fn host_info(&self) -> Option<Rc<RefCell<dyn Any>>> {
```
Ideally we'd move this out of the `wasmtime` crate entirely. Can all the `host_info` stuff move to `wasm_store_t` in the C API? I think the C API could store a `Rc<wasm_store_t>` on all the exported types, right?

fitzgen replied:
As mentioned in chat, it is useful to have this built into the `wasmtime` crate so that host infos can be cleaned up when their associated `externref` goes away.

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -2,224 +2,179 @@
+        let ExternRef { inner, store: _ } = self;
+        f.debug_struct("ExternRef")
+            .field("inner", &inner)
+            .field("store", &"..")
```
Could this pass in `self.store()` and defer to `Debug for Store`?

fitzgen replied:
`Store` doesn't implement `Debug`, but I can add an impl.

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -2,224 +2,179 @@
 /// Represents a piece of data located in the host environment.
-pub struct HostRef<T>(Rc<RefCell<ContentBox<T>>>);
+#[derive(Debug)]
+pub struct HostRef<T>
```
Can this structure move to the C API?

alexcrichton commented on crates/wasmtime/src/ref.rs:
```diff
@@ -2,224 +2,179 @@
-    Other(OtherRef),
+pub struct ExternRef {
+    pub(crate) inner: VMExternRef,
+    pub(crate) store: Weak<crate::runtime::StoreInner>,
```
The only reason this is here is for the `host_info` APIs, right? If so I would personally prefer that we just jettison that support from the C API for now. I think we should find a better way for implementing that, if necessary, instead of altering the design of the `wasmtime` crate itself.

fitzgen commented:
Alright, I believe I've addressed all of your review feedback and discussed the cases where I haven't (host info).

Want to take another look @alexcrichton?
